lib/auth/export.go

```diff
@@ -0,0 +1,228 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"context"
+	"encoding/pem"
+	"strings"
+	"time"
+
+	apidefaults "github.com/gravitational/teleport/api/defaults"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/sshutils"
+	"github.com/gravitational/trace"
+)
+
+// ExportAuthoritiesRequest has the required fields to create an export authorities request.
+// An empty AuthType exports all types.
+type ExportAuthoritiesRequest struct {
+	AuthType                   string
+	Client                     ClientI
+	ExportAuthorityFingerprint string
+	ExportPrivateKeys          bool
+	CompatVersion              string
+}
+
+// ExportAuthorities returns the list of authorities in OpenSSH compatible formats
+// If the ExportAuthoritiesRequest.AuthType is present only prints keys for CAs of this type,
+// otherwise returns all keys concatenated
+func ExportAuthorities(ctx context.Context, req ExportAuthoritiesRequest) (string, error) {
+	ret := strings.Builder{}
+
+	var typesToExport []types.CertAuthType
+
+	// this means to export TLS authority
+	switch req.AuthType {
+	// "tls" is supported for backwards compatibility.
+	// "tls-host" and "tls-user" were added later to allow export of the user
+	// TLS CA.
+	case "tls", "tls-host":
+		return exportTLSAuthority(ctx, req.Client, types.HostCA, false, req.ExportPrivateKeys)
+	case "tls-user":
+		return exportTLSAuthority(ctx, req.Client, types.UserCA, false, req.ExportPrivateKeys)
+	case "db":
+		return exportTLSAuthority(ctx, req.Client, types.DatabaseCA, false, req.ExportPrivateKeys)
+	case "tls-user-der", "windows":
+		return exportTLSAuthority(ctx, req.Client, types.UserCA, true, req.ExportPrivateKeys)
+	}
+
+	// if no --type flag is given, export HostCA and UserCA.
+	if req.AuthType == "" {
+		typesToExport = []types.CertAuthType{types.HostCA, types.UserCA}
+	} else {
+		authType := types.CertAuthType(req.AuthType)
+		if err := authType.Check(); err != nil {
+			return "", trace.Wrap(err)
+		}
+		typesToExport = []types.CertAuthType{authType}
+	}
+	localAuthName, err := req.Client.GetDomainName(ctx)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	// fetch authorities via auth API (and only take local CAs, ignoring
+	// trusted ones)
+	var authorities []types.CertAuthority
+	for _, at := range typesToExport {
+		cas, err := req.Client.GetCertAuthorities(ctx, at, req.ExportPrivateKeys)
+		if err != nil {
+			return "", trace.Wrap(err)
+		}
+		for _, ca := range cas {
+			if ca.GetClusterName() == localAuthName {
+				authorities = append(authorities, ca)
+			}
+		}
+	}
+
+	for _, ca := range authorities {
+		if req.ExportPrivateKeys {
```

Comment:
This is a pretty serious action, which safeguards do we have in place to make sure only the most trusted admins can get here?

Is it really, _really_, REALLY, necessary to allow exporting private keys? Could we scratch this feature altogether?

Reply:
The new endpoint calls the `client.ExportAuthorities` passing a hard coded `false` in the `ExportPrivateKeys` request.

Exporting private keys is only accessible via `tctl` and uses the same code path as before.
I didn't explore the entire code path, but the changes should not change the way it works.

Reply:
Not sure what safeguards we could put here. The only thing which comes to my mind is defining a separate function in `lib/client/ca_export.go` which doesn't accept `ExportPrivateKeys` as a param of any kind and always exports only public keys. Places such as `web.Handler.authExportPublic` could call that function instead. Hopefully that would be a little less prone to human error.

Reply:
+1 for Rafal's idea. tctl can call both functions if the appropriate flag is set.

Reply:
:heavy_check_mark: